File: ltms/go/indirect.go
// Indirect proof mechanism for LTRE.
// Converted from indirect.lisp.
package ltms

import "fmt"

// TryIndirectProof attempts to prove a fact by contradiction.
func TryIndirectProof(fact interface{}, ltre *LTRE) bool {
	if ltre == nil {
		ltre = CurrentLTRE
	}
	if IsKnown(fact, ltre) {
		return true
	}

	WithContradictionHandler(ltre.LTMS,
		func(contradictions []*Clause, l *LTMS) bool {
			assumptions := AssumptionsOfClause(contradictions[0])
			factDatum := Referent(fact, true, ltre)
			factNode := factDatum.TmsNode
			var theNode *TmsNode
			for _, asn := range assumptions {
				if asn == factNode {
					theNode = asn
					break
				}
			}
			if theNode != nil {
				status := theNode.Label
				RetractAssumption(theNode)
				AddNogood(theNode, status, assumptions)
				return true
			}
			return false
		},
		func() {
			Assuming([]interface{}{[]interface{}{":NOT", fact}}, ltre, func() {
				RunRules(ltre)
			})
		})

	return IsKnown(fact, ltre)
}

// IndirectProofExample demonstrates indirect proof.
func IndirectProofExample() {
	InLTRE(CreateLTRE("Indirect Proof Example", false))
	Assert([]interface{}{":OR", "p", "q"}, "user", CurrentLTRE)
	Assert([]interface{}{":IMPLIES", "p", "r"}, "user", CurrentLTRE)
	Assert([]interface{}{":IMPLIES", "q", "r"}, "user", CurrentLTRE)
	fmt.Printf("\nBefore indirect proof: r known? %v", IsKnown("r", CurrentLTRE))
	result := TryIndirectProof("r", CurrentLTRE)
	fmt.Printf("\nAfter indirect proof: r known? %v (result=%v)", IsKnown("r", CurrentLTRE), result)
}

File: ltms/cpp/indirect.cpp
/// Indirect Proof Mechanism - Implementation
/// Converted from indirect.lisp
/// Copyright (c) 1986-1993, Kenneth D. Forbus, Northwestern University,
/// and Johan de Kleer, the Xerox Corporation. All rights reserved.

#include "indirect.h"
#include "funify.h"
#include <iostream>
#include <algorithm>

// ================================================================
// try_indirect_proof
// ================================================================
// Attempts to prove a fact by assuming its negation and deriving
// a contradiction. If the negation leads to contradiction, the
// fact must be true.

bool try_indirect_proof(const std::any& fact, LTRE* ltre) {
    if (!ltre) ltre = current_ltre;

    if (is_known(fact, ltre)) return true;

    std::string fact_str = any_to_string(fact);

    with_contradiction_handler(ltre->ltms,
        [&](std::vector<Clause*>& contradictions, LTMS* ltms) -> bool {
            if (contradictions.empty()) return false;

            auto assumptions = assumptions_of_clause(contradictions[0]);
            Datum* datum = referent(fact, true, ltre);
            TmsNode* fact_node = datum->tms_node;

            auto it = std::find(assumptions.begin(), assumptions.end(), fact_node);
            if (it != assumptions.end()) {
                NodeLabel status = fact_node->label;
                retract_assumption(fact_node);

                // Build nogood from remaining assumptions
                std::vector<TmsNode*> remaining;
                for (auto* a : assumptions) {
                    remaining.push_back(a);
                }
                add_nogood(fact_node, status, remaining);
                return true;
            }
            return false;
        },
        [&]() {
            // Assume the negation of fact
            std::string neg_fact = "(:NOT " + fact_str + ")";

            Datum* datum = referent(std::any(neg_fact), true, ltre);
            TmsNode* node = datum->tms_node;
            convert_to_assumption(node);
            enable_assumption(node, NodeLabel::FALSE);

            run_rules(ltre);
        }
    );

    return is_known(fact, ltre);
}

// ================================================================
// Example
// ================================================================

void indirect_proof_example() {
    in_ltre(create_ltre("Indirect Proof Example"));

    // assert (:OR p q)
    ltre_assert(std::any(std::string("(:OR p q)")),
                std::any(std::string("user")));

    // assert (:IMPLIES p r)
    ltre_assert(std::any(std::string("(:IMPLIES p r)")),
                std::any(std::string("user")));

    // assert (:IMPLIES q r)
    ltre_assert(std::any(std::string("(:IMPLIES q r)")),
                std::any(std::string("user")));

    std::cout << "Before indirect proof, r known? "
              << (is_known(std::any(std::string("r"))) ? "yes" : "no")
              << std::endl;

    try_indirect_proof(std::any(std::string("r")));

    std::cout << "After indirect proof, r known? "
              << (is_known(std::any(std::string("r"))) ? "yes" : "no")
              << std::endl;
}
